### ISSUE TITLE
Improve compile-time flags

### ISSUE THREAD

**Issue by ysbaddaden** [crystal-lang/crystal#15157]
I miss a couple features related to compile time flags in the compiler:

1. **flags with a value**

   It can be emulated with `-Dkey=value` then checking for full `"key=value"` flags, for example `{% if flag?("evloop=epoll") %}`, it's a bit clunky but it works... until we need to manipulate the value(s), for example `-Dbranch-protection=bti,pac-ret,leaf` where we'd have to split on the comma to enable this or that feature in that specific order.

   For this we need support from the compiler. For example a `Crystal::Program#flag?(name)` method that would in order: check if a value-less flag exists, check if a value flag exists and return its value (string) or return false. Then we could update the `flag?` macro method to use it instead of `Crystal::Program#has_flag?`.

   I don't see any backward compatibility issues?

2. **defining flags from macros**

   Sometimes it would be useful to define a flag from macros, for example to set some default flags on each platform, right from the stdlib code, instead of hardcoding those in the compiler or having to duplicate the target checks in many places...

   There could for example be a pair of `set_flag(name)` and `set_flag(name, value)` macro methods.

**Comment by straight-shoota** [comment]
For retrieveing a value we have two options:

* Integrate into the existing `flag?` macro method. Maybe a trailing `=` could indicate that we want to get the value of the flag, not check its existence: `{{ flag?("branch-protection=") }} # => "bti,pac-ret,leaf"`. This would change the return type of `flag?` to `BoolLiteral | StringLiteral`. I don't see any issues with backward compatibility, but it might be piling a bit much on a single macro method.
* Add a new macro method `flag_value` (or `flag_option` or other). This would be a separate feature, clearly distinct from the `flag?` macro: `{{ flag_value("branch-protection") }} # => "bti,pac-ret,leaf"`. The return type would be `StringLiteral | NilLiteral`.

We also need to consider what happens when a flag key is passed multiple times: `-Dbranch-protection=bti Dbranch-protection=pac-ret`. What's the value of that? Does only the last one count? Or do we merge them together, as `StringLiteral` or `ArrayLiteral`? I'm not sure about concrete use cases, but it might be realistic that you might want to accept multiple values via individual assignments 🤷
This isn't relevant for `flag?` because setting a flag multiple times has the same effect as setting it once.

**Comment by straight-shoota** [comment]
I'm not sure about how useful defining flags from use code would be. A practical concern would be that this means flags change during the compilation process which can lead to unexpected behaviour.

We might need to look a bit into use cases to find a good motivation for this. But I suppose we can always emulate a similar effect by checking the definition of constants (as we did in #14996).

IMO flags derived from the target triple fit perfectly into the compiler. I don't see any benefit of moving them to code. It'll reduce performance (macros are quite fast, but native code is faster).

**Comment by ysbaddaden** [comment]
I was thinking more in terms of runtime features that may depend on some flag, for which we'd like to define default values depending on the target... or not (e.g. always enable a feature in a shard).

The only way to achieve that is to hardcode them in the compiler (and the default be unavailable until the next release), while a `set_flag` macro method wouldn't and allow apps and shards to enable features right in code (no need for cli flags):

```crystal
# always enable feature X because we need it
{% set_flag("mylib_enable_x") %}
require "mylib"
```

Checking on constants works but it feels clunky (we may not have different constants) and setting a flag should be _much_ faster than injecting a constant just for enabling some feature.

**Comment by ysbaddaden** [comment]
**A 3rd feature** would be to define compile time flags somewhere and have them applied by crystal automatically (no need for manual `-D` command line arguments). For example a `flags:` list in `shard.yml`.

I.e. this is reminiscent of [crystal editions](https://forum.crystal-lang.org/t/ensuring-smooth-language-changes-for-2-0/3588/14).

I was thinking about it as a counter argument to the set flag macro a bit, but I still find value in setting flags based upon other flags purely in code.

**Comment by straight-shoota** [comment]
Yeah flags for runtime features should not be defined by the compiler 👍 

I think it's hard to reason about the benefits of different approaches without concrete use cases.

For flag values we have a concrete use case, passing multiple `branch-protection` options.
But even that isn't really critical. It's already possible to use a couple of individual flags `-Dbranch-protection=bti -Dbranch-protection=pac-ret -Dbranch-protection=leaf` instead of a combined `-Dbranch-protection=bti,pac-ret,leaf`. It's a bit more verbose, but that's it. No need for a new compiler feature.

An explicit way to retrieve flag values is only really necessary when the possible values are unknown.
That would be a similar use case as environment flags, which we're already using for some compiler configuration (`CRYSTAL_CONFIG_CC`, `CRYSTAL_CONFIG_PATH`, `CRYSTAL_CONFIG_TARGET` etc.). But I don't think it makes much sense to turn them into compiler flags.

**Comment by straight-shoota** [comment]
I encountered another use case where more flexible compiler flags would be useful.

The target triple for some operating systems encodes an OS version identifier (e.g. `x86-64-freebsd14.0`). It would be nice if we could query that version identifier and use it with `compare_versions`.

Example use case:
https://github.com/crystal-lang/crystal/blob/67e3ad9901669f3fffb56fd42972f09e81fe0c16/src/lib_c/x86_64-freebsd/c/unistd.cr#L23-L25

The current implementation is insufficient. It ignores all minor releases between 12.0 and 14.0.

We could encode the version in an auto-generated compiler flag with version as value: `freebsd_version=14.0`
And then we could change the above code to this:
```cr
{% if compare_versions("#{(flag?(:freebsd_version) || "13.0").id}.0", "14.0.0") >= 0 %}
  fun execvpe(file : Char*, argv : Char**, envp : Char**) : Int 
{% end %}
```
Note that the fallback clause (`|| "13.0"`) makes this entirely compatible with older compiler versions where `flag?(:freebsd_version)` would return `false`.
The padding with a trailing `.0` is necessary for `compare_versions` to work with the two-component version numbers for FreeBSD releases.

**Comment by Sija** [comment]
> The padding with a trailing `.0` is necessary for `compare_versions` to work with the two-component version numbers for FreeBSD releases.

Perhaps `compare_versions` could accept shorter version strings to make things easier?

**Comment by straight-shoota** [comment]
@Sija Yes, I've been complaining about that for 6 years: https://github.com/crystal-lang/crystal/issues/7637 😆
But that's unrelated to the discussion here. I would've liked to exclude it, but considered it better to show a complete but working example, even with the additional complexity.

**Comment by ysbaddaden** [comment]
- We can consider that flags already have a `true` value by default, they'd now have a `StringLiteral` value. No breaking change.

- For backward compatibility `name=value` would create two flags:

  ```crystal
  {{ flag?(:name) }}         # => "value"
  {{ flag?(:"name=value") }} # => true
  ```

- Setting a flag twice should just replace the value IMO. Multiple values can be handled by the macros that need it, with a recommendation to split on `commas`:

  ```crystal
   {% if values = (x = flag?(:"evloop")) && x.split(",") %}
     {% for name in values %}
       require "./{{ name.id }}"
     {% end %}
   {% end %}
  ```

- There could be a general `abi_version` since the same logic applies to many targets (Android, *BSD, Darwin, ...):

  ```crystal
  {% if flag?(:freebsd) &&
       compare_versions(flag?(:abi_version), "14.0") >= 0) %}
   ```

**Comment by straight-shoota** [comment]
Turns out, the freebsd version flag only represents the major version. So it's not `freebsd12.0` but `freebsd12`.
This is much easier to use and the specific use case from https://github.com/crystal-lang/crystal/issues/15157#issuecomment-3476932528 should be fine without flag values (see #16311).

**Comment by straight-shoota** [comment]
I think it's time to clean up a bit and close down this issue. It has served its main purpose.

1. Flag values has been implemented in https://github.com/crystal-lang/crystal/pull/16310
2. Defining flags from macros still seems a bit pale. I don't think we have explored practical use cases and implications deep enough. The discussion here has focused primarily on other aspects. This should probably warrant a separate issue.
3. Defining flags somewhere in a file like `shard.yml` is no more than a rough idea that could lead to a feature request for the compiler or shards, or perhaps just using better build tooling. This certainly needs some more exploration first, which we might start in a forum thread maybe?
4. Deriving implicit value flags from target properties: #17373
